crates/rolldown/src/utils/fold_const_value.rs

```diff
@@ -0,0 +1,51 @@
+use oxc::allocator::Allocator;
+use oxc::ast::{ast::*, visit::walk_mut, AstBuilder, VisitMut};
+use oxc::span::SPAN;
+use oxc::syntax::operator::BinaryOperator;
+
+/// This is the very basic version of Inline Binary, aiming to support `process.env.NODE_ENV === "production"` -> `true` or `false`.
+/// Used before dead branch elimination.
+pub struct InlineBinaryValueLight<'a> {
```

Comment:
Oxc is working on this: https://github.com/oxc-project/oxc/blob/main/crates/oxc_minifier/src/ast_passes/replace_global_defines.rs

I think is best we improve oxc rather than working on an temparay solution.

Reply:
I think it is related to https://github.com/rolldown/rolldown/issues/186, rather than `BinaryExpr` inline

Reply:
I got it wrong. Then I think you should improve https://github.com/oxc-project/oxc/blob/main/crates/oxc_minifier/src/ast_passes/remove_dead_code.rs. @Boshen cc

Feel like duplicate works of oxc. If we are using oxc's `Compressor`, this feature should be provided.

Reply:
From my understanding, boshen don't want rolldown use `Compressor` to do such things, that's why https://github.com/oxc-project/oxc/issues/3706 get closed and `RemoveDeadCode` is introduced.

But `RemoveDeadCode` is immature now, so to do such things, I guess we need port some features of `Compressor` to `RemoveDeadCode`.

Reply:
The full implementation for constant folding is not trivial (And we don't need all the features of  it) . Also It may hurt performance in the scan phase, or oxc needs to implement a lightweight version of const folding, which is not necessary. The current implementation is enough for most cases, and why I mark this as a temporary solution is that I want to find a solution to merge all the preprocessor visitors in one go, which should be a parser hook or something else.

Reply:
> From my understanding, boshen don't want rolldown use `Compressor` to do such things, that's why [oxc-project/oxc#3706](https://github.com/oxc-project/oxc/issues/3706) get closed and `RemoveDeadCode` is introduced.
> 
> But `RemoveDeadCode` is immature so far, so to do such things, I guess we need port some features of `Compressor` to `RemoveDeadCode`.

https://github.com/oxc-project/oxc/blob/main/crates/oxc_minifier/src/compressor/fold.rs, this is what we want.

Reply:
But it looks overkill for supporting `process.env.NODE_ENV === "production"`  and dead branch eliminating

Reply:
> why I mark this as a temporary solution is that I want to find a solution to merge all the preprocessor visitors in one go

Let me put it this way. We weren't urgent to support this feature. So let's have a discussion with @Boshen and try find a globally optimal solution.

Personally, I have questions:

- If we should maintain this logic in rolldown's repo, or should we do it in oxc's repo?
- Could/Should we reuse oxc's logic?

---

> The current implementation is enough for most cases

I don't against merging this PR, since it's almost finished. But I hope we won't stop there, since this is only temporary solution as you said.

Reply:
> But it looks overkill for supporting `process.env.NODE_ENV === "production"` and dead branch eliminating

Not sure. For example I'm thinking about how we support `require('./' + 'foo' + ".js")`. This could be done by the complete version of folding expression.

And

```js
if (sss || 'foo') {
  ...
}
```

should become

```js
{
  ...
}
```

We probably gonna need the complete version of folding expression.